Accept service id in reports

davidor opened the pull request:
Closes https://github.com/3scale/3scale_ws_api_for_ruby/issues/29

This breaks compatibility with previous versions, as it introduces a new parameter in `ThreeScale::Client#report`. If this is merged, we should release a new major version.

mikz commented on lib/3scale/client.rb:
```diff
@@ -122,20 +124,27 @@ def authrep(options)
+    #    client.report(transactions: [{:app_id => 'foo', :usage => {'hits' => 1}}],
+    #                  service_id: 'a_service_id')
+    #
+    def report(transactions: [], service_id: nil)
```
There is a way to do this in backwards compatible manner no?

```ruby
def report(*transactions)
  reports, service_id = transactions
  reports << service_id unless service_id.is_a?(Numberic)
  ...
end
```

davidor replied:
That would not work, because of the '*' in transactions.
We could do something like `reports = transactions[0..-2]`, and then see if the last element in transactions is a hash(transaction) or Numeric/String (service_id).

davidor replied:
That would make the code uglier and the interface less intuitive, but we would not break the compatibility with previous versions. I see that we are not encouraging locking to a major version in the readme: `gem '3scale_client'` so it might be a problem for some users...

I'm thinking that it might be better to do this in a backwards compatible way. What do you think @mayorova ? @mpguerra ?

mikz replied:
I'd be pissed as a developer to get this kind of breaking change without deprecation warning. We use this code in system. 

Updated snippet:
```ruby
def report(*transactions)
  *reports, service_id = transactions
  unless service_id.is_a?(Numeric)
    reports << service_id
    service_id = nil
  end
  ...
end
```

mikz replied:
That would be breaking change I'd be okay with. Just drop <= 2.0 compatibility.

mayorova replied:
Some comments: 

1) Agree with @mikz that the best way would be to keep the solution backward compatible, but add a deprecation warning and encourage everybody use the `(transactions: [], service_id: 123)` interface.

2) The solution:

    def report(*reports, transactions: [], service_id: nil)
        ...
    end

doesn't work for the previous interface, for example: `report({app_id: 1, usage: 2}, {app_id: 3, usage: 4})`.
It fails with `ArgumentError: unknown keywords: app_id, usage`, other than that a really nice solution

3) Regarding the solution:
    def report(*transactions)
        *reports, service_id = transactions
        unless service_id.is_a?(Numeric)
        ...
    end

To be honest I don't like the interface too much (an optional numeric value as the last element of the list).

4) An alternative solution:

    def report(*params)
      if params.first.has_key?(:transactions)
        transactions = params.first[:transactions]
        service_id = params.first[:service_id]
      else
        transactions = params
        warn 'deprecation warning'
      end
      # here the `transactions` has the array of reports
      ...
    end

Not very elegant... But it will work with all the following:


    report({app_id: 1, usage: 2}, {app_id: 3, usage: 4})
    report({app_id: 1, usage:2})
    report({transactions: [{app_id: 1, usage: 2}, {app_id: 3, usage: 4}], service_id: 111})
    report(transactions: [{app_id: 1, usage: 2}, {app_id: 3, usage: 4}])

5) I don't know if there is a problem to drop <= 2.0 compatibility... We don't know if any of the existing customers user it.

mikz replied:
Ok. Point taken. Updated solution:

```ruby
def report(*reports, transactions: [], service_id: nil, **rest)
  transactions.concat(reports).concat([rest])
  ...
end
```

mayorova replied:
Yep, this last one is cool! 

I tried with just `**rest`, but the first `*reports` was the missing part.

Would still be good to add the deprecation note somewhere and update the doc. Other than that I think this should be good.

davidor replied:
That code is a bit tricky, but it's looks like a good compromise to keep backwards compatibility :+1:

mayorova replied:
One thing to keep in mind is that `transactions.concat(reports).concat([rest])` may add an empty object to the `transactions` array (if `transactions: []` param is used), which will later break in `encode_transactions`.

So, either it should be filtered out before, or validated first inside `encode_transactions`.

mikz replied:
True! Very good point. It should concat only when `rest` is not empty!

davidor commented:
The build fails because I used a syntax that was introduced in Ruby 2.0: `def report(transactions: [], service_id: nil)` and we are testing with older Rubies too :smile: 
I will fix that.

davidor commented:
@mikz , @mayorova : I have changed the method as we agreed in the comments. I have also added a test that checks the backwards compatibility with the old report method: `test_report_compatibility_with_old_report_format`

davidor commented:
I'll drop the compatibility with Ruby < 2.0 in a different PR as we agreed. Travis is going to fail for Rubies < 2.0 in this PR.

davidor commented:
Wait. This fails if only one transaction is sent, using the old format, and with strings as the hash keys:
`{ 'app_id'    => 'app_id_1', 'usage'    => { 'hits' => 1 } }` . There's a bug in the condition that checks that there aren't any transactions to report.

mayorova commented:
I think using strings as hash keys was not supported anyway, was it?..
The PR looks good to me now! 👍

mikz commented:
This is great change done in very polite way 👍

davidor commented:
You are right @mayorova . I got confused.
This PR can be merged now.

mikz commented:
One last comment. This change definitely needs a CHANGELOG entry.

davidor commented:
Sure @mikz, I'll do that in the release commit when we change the version number.

mikz commented:
Ok. I thought we are going to accumulate more changes before the release.
Then we would them to `[unreleased]` like https://github.com/3scale/3scale_ws_api_for_ruby/blob/8cbbf8be3c06f20400c22c4c904c83bff24e354a/CHANGELOG.md and just change it to the release number when it actually gets released. It becomes harder over time to accumulate all notable changes.

unleashed commented:
We could adopt an automated changelog and versioning scheme like the well known one used in Angular. But in general, I feel it is part of the release engineer's work to produce this document, since it is their prerogative to include or not a change in a given release and highlight or not changes, including their own wording. At least this is the way traditional releng works (which does not mean the changelog should be left untouched until the last moment). As many other things, I like pointing out the git.git repo, where the maintainers keep a RelNotes.txt file regularly up-to-date.

davidor commented:
@mikz I have documented the changes of this PR in the changelog. We can use this approach from now on.

mikz commented:
@davidor awesome. Love it ❤️ 👍
